### PR TITLE
SONAR-6948 Ability for Java server extensions to call web services

### PR DESCRIPTION
**Pull request by simonbrandhof**

### CODE REVIEW

File: it/it-plugins/ws-plugin/src/main/java/LocalCallWebService.java [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+import org.sonarqube.ws.client.GetRequest;
+import org.sonarqube.ws.client.WsClient;
+import org.sonarqube.ws.client.WsClientFactory;
+import org.sonarqube.ws.client.WsResponse;
+
+public final class LocalCallWebService implements WebService {
+
+  @Override
+  public void define(Context context) {
+    NewController controller = context.createController("api/local_ws_call");
```
Comment (sns-seb) [inline it/it-plugins/ws-plugin/src/main/java/LocalCallWebService.java]: is it required to have `api/`?
could be interesting to have the IT not use it

Comment (simonbrandhof) [inline it/it-plugins/ws-plugin/src/main/java/LocalCallWebService.java]: well, it's the standard, but why not.

File: it/it-tests/src/test/java/it/ws/InterWsCallTest.java [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+    WsClient wsClient = WsClientFactory.getDefault().create(HttpConnector.builder().url(orchestrator.getServer().getUrl()).build());
+    WsResponse response = wsClient.wsConnector().call(new GetRequest("api/local_ws_call/protobuf_data"));
+    assertThat(response.isSuccessful()).isTrue();
+  }
+
+  @Test
+  public void gets_json() {
+    WsClient wsClient = WsClientFactory.getDefault().create(HttpConnector.builder().url(orchestrator.getServer().getUrl()).build());
+    WsResponse response = wsClient.wsConnector().call(new GetRequest("api/local_ws_call/json_data"));
+    assertThat(response.code()).isEqualTo(200);
```
Comment (sns-seb) [inline it/it-tests/src/test/java/it/ws/InterWsCallTest.java]: why use `code()` here and below but `isSuccessful()` above?

File: sonar-plugin-api/src/main/java/org/sonar/api/server/ws/LocalConnector.java [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+    String getMethod();
+
+    boolean hasParam(String key);
+
+    @CheckForNull
+    String getParam(String key);
+  }
+
+  interface LocalResponse {
+    int getStatus();
```
Comment (sns-seb) [inline sonar-plugin-api/src/main/java/org/sonar/api/server/ws/LocalConnector.java]: could document that it's the HTTP status code

File: sonar-plugin-api/src/main/java/org/sonar/api/server/ws/LocalConnector.java [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+
+    String getMediaType();
```
Comment (sns-seb) [inline sonar-plugin-api/src/main/java/org/sonar/api/server/ws/LocalConnector.java]: is it the mime type?

File: sonar-plugin-api/src/main/java/org/sonar/api/server/ws/LocalConnector.java [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+ * }
+ * </pre>
+ * </p>
+ * @since 5.5
+ */
+@Beta
+public interface LocalConnector {
+
+  interface LocalRequest {
+    String getPath();
```
Comment (sns-seb) [inline sonar-plugin-api/src/main/java/org/sonar/api/server/ws/LocalConnector.java]: could state that's it's the HTTP path

File: sonar-plugin-api/src/main/java/org/sonar/api/server/ws/LocalConnector.java [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+
+    String getMediaType();
```
Comment (sns-seb) [inline sonar-plugin-api/src/main/java/org/sonar/api/server/ws/LocalConnector.java]: is it the mime type?

File: sonar-plugin-api/src/main/java/org/sonar/api/server/ws/internal/SimpleGetRequest.java [hunk 7]
```diff
@@ -76,4 +78,16 @@ public SimpleGetRequest setParam(String key, @Nullable String value) {
     return this;
   }
 
+  public Map<String, String> getParams() {
+    return params;
+  }
+
+  @Override
+  public LocalConnector getLocalConnector() {
+    return localConnector;
```
Comment (sns-seb) [inline sonar-plugin-api/src/main/java/org/sonar/api/server/ws/internal/SimpleGetRequest.java]: either missing a `@CheckForNull` or some check in the implementation is missing to enforce the method's contract

File: sonar-plugin-api/src/main/java/org/sonar/api/server/ws/internal/SimpleGetRequest.java [hunk 8]
```diff
@@ -37,6 +38,7 @@
 
   private final Map<String, String> params = Maps.newHashMap();
   private String mediaType = "application/json";
+  private LocalConnector localConnector;
```
Comment (sns-seb) [inline sonar-plugin-api/src/main/java/org/sonar/api/server/ws/internal/SimpleGetRequest.java]: missing `@CheckForNull`

File: sonar-plugin-api/src/main/java/org/sonar/api/server/ws/internal/ValidatingRequest.java [hunk 9]
```diff
@@ -49,6 +51,15 @@ public void setAction(WebService.Action action) {
   }
 
   @Override
+  public LocalConnector getLocalConnector() {
```
Comment (sns-seb) [inline sonar-plugin-api/src/main/java/org/sonar/api/server/ws/internal/ValidatingRequest.java]: either missing a `@CheckForNull` or some check in the implementation is missing to enforce the method's contract

File: sonar-plugin-api/src/main/java/org/sonar/api/server/ws/internal/ValidatingRequest.java [hunk 10]
```diff
@@ -39,6 +40,7 @@
 public abstract class ValidatingRequest extends Request {
 
   private WebService.Action action;
+  private LocalConnector localConnector;
```
Comment (sns-seb) [inline sonar-plugin-api/src/main/java/org/sonar/api/server/ws/internal/ValidatingRequest.java]: missing `@CheckForNull`

File: sonar-plugin-api/src/main/java/org/sonar/api/server/ws/internal/ValidatingRequest.java [hunk 11]
```diff
@@ -49,6 +51,15 @@ public void setAction(WebService.Action action) {
   }
 
   @Override
+  public LocalConnector getLocalConnector() {
+    return localConnector;
+  }
+
+  public void setLocalConnector(LocalConnector lc) {
```
Comment (sns-seb) [inline sonar-plugin-api/src/main/java/org/sonar/api/server/ws/internal/ValidatingRequest.java]: nothing really prevents me from never calling this method or pass it `null` and break the contract of `getLocalConnector`

File: sonar-ws/src/main/java/org/sonarqube/ws/client/LocalConnectorAdapter.java [hunk 12]
```diff
@@ -0,0 +1,132 @@
...
+import com.google.common.annotations.VisibleForTesting;
+import java.io.ByteArrayInputStream;
+import java.io.InputStream;
+import java.io.InputStreamReader;
+import java.io.Reader;
+import org.sonar.api.server.ws.LocalConnector;
+
+import static java.nio.charset.StandardCharsets.UTF_8;
+
+class LocalConnectorAdapter implements WsConnector {
```
Comment (sns-seb) [inline sonar-ws/src/main/java/org/sonarqube/ws/client/LocalConnectorAdapter.java]: isn't this class a `LocalWsConnector`?

File: sonar-plugin-api/src/main/java/org/sonar/api/server/ws/LocalConnector.java (hunk 4 above)
Comment (simonbrandhof) [inline sonar-plugin-api/src/main/java/org/sonar/api/server/ws/LocalConnector.java]: yes, media type is the term used by sonar-ws and sonar-plugin-api.

Comment (sns-seb) [inline sonar-plugin-api/src/main/java/org/sonar/api/server/ws/LocalConnector.java]: it's worth documenting

File: sonar-ws/src/main/java/org/sonarqube/ws/client/WsClientFactory.java [hunk 13]
```diff
@@ -0,0 +1,41 @@
...
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonarqube.ws.client;
+
+import org.sonar.api.server.ws.LocalConnector;
+
+/**
+ * @since 5.5
+ */
+public class WsClientFactory {
```
Comment (sns-seb) [inline sonar-ws/src/main/java/org/sonarqube/ws/client/WsClientFactory.java]: this design closes the door to any testing and create a static binding :(

Comment (sns-seb) [inline sonar-ws/src/main/java/org/sonarqube/ws/client/WsClientFactory.java]: of course, usage of previous design in previous version was static too
the workaround would have been to use a property in the class and a constructor argument
because we choose to use a static binding in our implementations is not a reason to force other user to do the same, hence the initial proposed design of this class

Comment (sns-seb) [inline sonar-ws/src/main/java/org/sonarqube/ws/client/WsClientFactory.java]: in fact, we are not using the feature in our own code, but we should
taking `WsClientFactory` as an example, why aren't we adding `WsClientFactory` as a parameter of method `provide` and add `WsClientFactory.getDefault()` result to the container?
We lack time to provide a test implementation of `WsClientFactory` (which would, for example, allow to directly provide responses), it's not a good enough reason to implement a flexible and (imho) better design

Comment (sns-seb) [inline sonar-ws/src/main/java/org/sonarqube/ws/client/WsClientFactory.java]: I was giving thoughs on what pattern we should recommend to plugin developer to follow and I hit an issue with having a non static `WsClientFactory`.

I thought of advising the plugin developer to design, let's say `FooWs`, to get injected with a `WsClientFactory` instance and add `WsClientFactory.getDefault()` to the component list returned by its `FooPlugin` class. This way, the UTs can use a test implementation of `WsClientFactory` and he/she can easily emulates internal Ws calls in SQ.

Now, assume the pattern described above hass been implemented: we design our components to be injected with `WsClientFactory` and therefor SQ's WebServer's container contains a `WsClientFactory.getDefault()` object.

Add to this the hypothesis that the plugin does not bundle the same version of the `WsClient` library as SQ does.

What is going to be Pico's behavior here? We have 2 objects of the same class `WsClientFactory` but from different class loaders?

I really don't know.

I'm starting to doubt the library approach...

File: sonar-ws/src/test/java/org/sonarqube/ws/client/LocalWsConnectorTest.java [hunk 14]
```diff
@@ -0,0 +1,150 @@
...
+
+  @Test
+  public void baseUrl_is_always_slash() {
+    assertThat(underTest.baseUrl()).isEqualTo("/");
+  }
+
+  @Test
+  public void call_request() throws Exception {
+    WsRequest wsRequest = new PostRequest("api/issues/search")
+      .setMediaType(MediaTypes.JSON)
```
Comment (sns-seb) [inline sonar-ws/src/test/java/org/sonarqube/ws/client/LocalWsConnectorTest.java]: shouldn't `MediaType` be part of the API? This way it can be documented, users have a convenient way to test the value of `getMediaType()` without creating a constant of there own, etc.

File: sonar-plugin-api/src/main/java/org/sonar/api/server/ws/LocalConnector.java [hunk 15]
```diff
@@ -0,0 +1,110 @@
...
+    /**
+     * @see Request#getMediaType()
+     */
+    String getMediaType();
+
+    /**
+     * HTTP method. Possible values are "GET" and "POST"
+     * @see Request#method()
+     */
+    String getMethod();
```
Comment (sns-seb) [inline sonar-plugin-api/src/main/java/org/sonar/api/server/ws/LocalConnector.java]: shouldn't there be constants in the API representing the values `GET` and `POST` to avoid user define there own constants (or use magic strings in there code)?

Comment (sns-seb) [inline sonar-plugin-api/src/main/java/org/sonar/api/server/ws/LocalConnector.java]: this way, we could add a `@see`, for each value, to the Javadoc of this method

File: server/sonar-server/src/main/java/org/sonar/server/ws/WsUtils.java [hunk 16]
```diff
@@ -51,10 +52,12 @@ public static void writeProtobuf(Message msg, Request request, Response response
     OutputStream output = response.stream().output();
     try {
       if (request.getMediaType().equals(MediaTypes.PROTOBUF)) {
+        System.out.println("write protobuf " + Arrays.toString(msg.toByteArray()));
         response.stream().setMediaType(MediaTypes.PROTOBUF);
         msg.writeTo(output);
       } else {
         response.stream().setMediaType(MediaTypes.JSON);
+        System.out.println("write json ");
```
Comment (sns-seb) [inline server/sonar-server/src/main/java/org/sonar/server/ws/WsUtils.java]: this should be removed or replaced with a Logger

File: server/sonar-server/src/main/java/org/sonar/server/ws/WsUtils.java [hunk 17]
```diff
@@ -51,10 +52,12 @@ public static void writeProtobuf(Message msg, Request request, Response response
     OutputStream output = response.stream().output();
     try {
       if (request.getMediaType().equals(MediaTypes.PROTOBUF)) {
+        System.out.println("write protobuf " + Arrays.toString(msg.toByteArray()));
```
Comment (sns-seb) [inline server/sonar-server/src/main/java/org/sonar/server/ws/WsUtils.java]: this should be removed or replaced with a Logger

File: sonar-plugin-api/src/main/java/org/sonar/api/server/ws/LocalConnector.java (hunk 15 above)
Comment (simonbrandhof) [inline sonar-plugin-api/src/main/java/org/sonar/api/server/ws/LocalConnector.java]: nope, it's not needed as this API is used only by sonar-ws. Plugins do not access to the inner interfaces LocalRequest and LocalResponse

File: server/sonar-server/src/main/java/org/sonar/server/ws/WebServiceEngine.java [hunk 18]
```diff
@@ -112,9 +127,11 @@ public void execute(ValidatingRequest request, ServletResponse response,
     return action;
   }
 
-  private void sendErrors(ServletResponse response, int status, Errors errors) {
-    ServletResponse.ServletStream stream = response.stream();
-    stream.reset();
+  private void sendErrors(Response response, int status, Errors errors) {
+    Response.Stream stream = response.stream();
+    if (stream instanceof ServletResponse.ServletStream) {
```
Comment (sns-seb) [inline server/sonar-server/src/main/java/org/sonar/server/ws/WebServiceEngine.java]: I find it suspicious that reset must be called here
can you explain why it is required?

File: server/sonar-server/src/main/java/org/sonar/server/ce/ws/TaskTypesAction.java [hunk 19]
```diff
@@ -57,6 +58,7 @@ public void handle(Request request, Response response) throws Exception {
       .addAllTaskTypes(taskTypes)
       .build();
 
+    System.out.println("WS writes protobuf-----" + taskTypesWsResponse.getSerializedSize() + ":" + Arrays.toString(taskTypesWsResponse.toByteArray()));
```
Comment (sns-seb) [inline server/sonar-server/src/main/java/org/sonar/server/ce/ws/TaskTypesAction.java]: this should be removed

File: sonar-ws/src/test/java/org/sonarqube/ws/client/LocalWsConnectorTest.java (hunk 14 above)
Comment (simonbrandhof) [inline sonar-ws/src/test/java/org/sonarqube/ws/client/LocalWsConnectorTest.java]: You mix the scope of this PR with all the improvements that can be done on the WS extension point !

File: sonar-ws/src/main/java/org/sonarqube/ws/client/BaseService.java [hunk 20]
```diff
@@ -56,8 +58,10 @@ protected WsResponse call(WsRequest request) {
 
   public <T extends Message> T convert(WsResponse response, Parser<T> parser) {
     try (InputStream byteStream = response.contentStream()) {
+      byte[] bytes = IOUtils.toByteArray(byteStream);
+      System.out.println("BaseService.convert " + bytes.length + " bytes : " + Arrays.toString(bytes));
```
Comment (sns-seb) [inline sonar-ws/src/main/java/org/sonarqube/ws/client/BaseService.java]: any reason to change this beside logging?
if not, modifications to this class should be removed

File: sonar-ws/src/main/java/org/sonarqube/ws/client/HttpConnector.java [hunk 21]
```diff
@@ -192,16 +192,20 @@ private WsResponse post(PostRequest postRequest) {
...
     try {
       Response okResponse = call.execute();
-      return new HttpResponse(okResponse);
+      return new OkHttpResponse(okResponse);
     } catch (IOException e) {
       throw new IllegalStateException("Fail to request " + okRequest.urlString(), e);
     }
   }
 
+  public static Builder builder() {
```
Comment (sns-seb) [inline sonar-ws/src/main/java/org/sonarqube/ws/client/HttpConnector.java]: if you provide a static factory method to create a builder:

- it would remove an ambiguity to call it `newBuilder` (returns a new instance for each call)
- constructor of Builder class should be private to enforce use of the static factory method

File: sonar-ws/src/main/java/org/sonarqube/ws/client/WsClientFactory.java (hunk 13 above)
Comment (simonbrandhof) [inline sonar-ws/src/main/java/org/sonarqube/ws/client/WsClientFactory.java]: I don't agree on the need. sonar-ws is a library and by design it can't be provided at runtime as sonar-plugin-api. The constraints are not the same at all. In this context having `WsClientFactory.getDefault().newClient(HttpConnector.builder()....build())` is too complex. Let's keep things simple and do not overthink about future needs.

Comment (simonbrandhof) [inline sonar-ws/src/main/java/org/sonarqube/ws/client/WsClientFactory.java]: An application using the HTTP client has two ways to test:
1. implementing its own own WsClient (which is an interface). Whatever the way to create an instance (Factory with static methods or others), it's up to the application to decouple the usage of WsClient with the way to instantiate it.
2. using a real HTTP server

In the context of SQ plugins defining new web services, `WsTester` should be improved in order to provide the ability to emulate interactions with other WS.

File: server/sonar-server/src/main/java/org/sonar/server/ws/WebServiceEngine.java (hunk 18 above)
Comment (simonbrandhof) [inline server/sonar-server/src/main/java/org/sonar/server/ws/WebServiceEngine.java]: this is out of scope of this PR.

File: sonar-ws/src/main/java/org/sonarqube/ws/client/HttpConnector.java (hunk 21 above)
Comment (simonbrandhof) [inline sonar-ws/src/main/java/org/sonarqube/ws/client/HttpConnector.java]: the constructor was kept public for backward-compatibility, but it's not a good reason.

File: sonar-plugin-api/src/main/java/org/sonar/api/server/ws/LocalConnector.java [hunk 22]
```diff
@@ -0,0 +1,110 @@
...
+
+    /**
+     * @see Response#setHeader(String, String)
+     * @return
+     */
+    @CheckForNull
+    String getHeader(String name);
+  }
+
+  LocalResponse call(LocalRequest request);
```
Comment (sns-seb) [inline sonar-plugin-api/src/main/java/org/sonar/api/server/ws/LocalConnector.java]: I'd put this method before the inner interface
each time I've read this interface, I've wondered what it defines

File: sonar-plugin-api/src/main/java/org/sonar/api/server/ws/internal/ValidatingRequest.java (hunk 9 above)
Comment (sns-seb) [inline sonar-plugin-api/src/main/java/org/sonar/api/server/ws/internal/ValidatingRequest.java]: what about that point, @simonbrandhof ?

Do you think we can ignore it because, even though this class is in the API, it is in an internal package?

File: sonar-ws/src/main/java/org/sonarqube/ws/client/DefaultWsClientFactory.java [hunk 23]
```diff
@@ -0,0 +1,25 @@
...
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonarqube.ws.client;
+
+interface DefaultWsClientFactory extends WsClientFactory {
```
Comment (sns-seb) [inline sonar-ws/src/main/java/org/sonarqube/ws/client/DefaultWsClientFactory.java]: is this interface still needed?

File: sonar-plugin-api/src/main/java/org/sonar/api/server/ws/internal/ValidatingRequest.java (hunk 11 above)
Comment (simonbrandhof) [inline sonar-plugin-api/src/main/java/org/sonar/api/server/ws/internal/ValidatingRequest.java]: Return value can't be null:
* ValidatingRequest checks that it is not null
* SimpleGetRequest throws UnsupportedOperationException

File: sonar-plugin-api/src/main/java/org/sonar/api/server/ws/LocalConnector.java [hunk 24]
```diff
@@ -0,0 +1,78 @@
...
+/**
+ * This class allows a web service to call another web service through the sonar-ws library.
+ * The call is in-process, synchronous and does not involve the HTTP stack.
+ * <p>
+ * Example of a web service that uses sonar-ws 5.5 to get some issues:
+ * <pre>
+ * public class MyRequestHandler implements org.sonar.api.server.ws.RequestHandler {
+ *   {@literal @}Override
+ *   public void handle(Request request, Response response) {
+ *     WsClient wsClient = WsClientFactory.getDefault().create(request.getLocalConnector());
```
Comment (simonbrandhof) [inline sonar-plugin-api/src/main/java/org/sonar/api/server/ws/LocalConnector.java]: Javadoc already fixed.

**Comment by sns-seb** [comment]
some minor comments left, otherwise LGTM (need a rebase on master, though, and building QA tests on Travis)

File: sonar-ws/src/main/java/org/sonarqube/ws/client/DefaultWsClientFactory.java (hunk 23 above)
Comment (simonbrandhof) [inline sonar-ws/src/main/java/org/sonarqube/ws/client/DefaultWsClientFactory.java]: exact, it's being removed.

**Comment by SonarTech** [comment]
#### Analysis summary
SonarQube analysis reported 5 issues

#### 5 unreported issues

Note: the following issues could not be reported as comments because they are located on lines that are not displayed in this pull request:

1. [Request.java#L212](https://github.com/SonarSource/sonarqube/blob/e8a84edfb18670fc511b5634d1ecbe39ea6eee95/sonar-plugin-api/src/main/java/org/sonar/api/server/ws/Request.java#L212): CRITICAL Either log or rethrow this exception. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1166)
1. [WebServiceEngine.java#L105](https://github.com/SonarSource/sonarqube/blob/e8a84edfb18670fc511b5634d1ecbe39ea6eee95/server/sonar-server/src/main/java/org/sonar/server/ws/WebServiceEngine.java#L105): CRITICAL Either log or rethrow this exception. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1166)
1. [WebServiceEngine.java#L107](https://github.com/SonarSource/sonarqube/blob/e8a84edfb18670fc511b5634d1ecbe39ea6eee95/server/sonar-server/src/main/java/org/sonar/server/ws/WebServiceEngine.java#L107): CRITICAL Either log or rethrow this exception. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1166)
1. org.sonarsource.sonarqube:sonar-process:target/generated-sources/protobuf/org/sonar/process/systeminfo/protobuf: MAJOR Add a 'package-info.java' file to document the 'protobuf' package [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1228)
1. [Request.java#L213](https://github.com/SonarSource/sonarqube/blob/e8a84edfb18670fc511b5634d1ecbe39ea6eee95/sonar-plugin-api/src/main/java/org/sonar/api/server/ws/Request.java#L213): MINOR Remove this use of "SonarException"; it is deprecated. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3ACallToDeprecatedMethod)
